test/jdk/jdk/jfr/event/io/TestSocketAdapterEvents.java

```diff
@@ -0,0 +1,126 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.jfr.event.io;
+
+import static jdk.test.lib.Asserts.assertEquals;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.nio.ByteBuffer;
+import java.nio.channels.ServerSocketChannel;
+import java.nio.channels.SocketChannel;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+
+import jdk.jfr.Recording;
+import jdk.jfr.consumer.RecordedEvent;
+import jdk.test.lib.jfr.Events;
+import jdk.test.lib.thread.TestThread;
+import jdk.test.lib.thread.XRun;
+
+/**
+ * @test
+ * @key jfr
+ * @requires vm.hasJFR
+ * @library /test/lib /test/jdk
+ * @run main/othervm jdk.jfr.event.io.TestSocketAdapterEvents
+ */
+public class TestSocketAdapterEvents {
+    private static final int writeInt = 'A';
+    private static final byte[] writeBuf = { 'B', 'C', 'D', 'E' };
+
+    private List<IOEvent> expectedEvents = new ArrayList<>();
+
+    private synchronized void addExpectedEvent(IOEvent event) {
+        expectedEvents.add(event);
+    }
+
+    public static void main(String[] args) throws Throwable {
+        new TestSocketAdapterEvents().test();
+    }
+
+    public void test() throws Throwable {
+        try (Recording recording = new Recording()) {
+            try (ServerSocketChannel ssc = ServerSocketChannel.open()) {
+                recording.enable(IOEvent.EVENT_SOCKET_READ).withThreshold(Duration.ofMillis(0));
+                recording.enable(IOEvent.EVENT_SOCKET_WRITE).withThreshold(Duration.ofMillis(0));
+                recording.start();
+
+                ssc.socket().setReuseAddress(true);
+                InetAddress lb = InetAddress.getLoopbackAddress();
+                ssc.bind(new InetSocketAddress(lb, 0));
+
+                TestThread readerThread = new TestThread(new XRun() {
+                    @Override
+                    public void xrun() throws IOException {
+                        byte[] bs = new byte[4];
+                        try (SocketChannel sc = ssc.accept(); Socket s = sc.socket();
+                             InputStream is = s.getInputStream()) {
+
+                            int readInt = is.read();
+                            assertEquals(readInt, writeInt, "Wrong readInt");
+                            addExpectedEvent(IOEvent.createSocketReadEvent(1, s));
+
+                            int bytesRead = is.read(bs, 0, 3);
+                            assertEquals(bytesRead, 3, "Wrong bytesRead partial buffer");
```

Comment:
This is probably okay although in theory the read method may return 1 or 2 bytes here. Initially I thought you should be using readNBytes here but that may involve more than one read and thus disrupt the number of events expected.